packages/firestore/exp/src/api/reference.ts

```diff
@@ -42,6 +53,131 @@ export function getDoc<T>(
   });
 }
 
+export function setDoc<T>(
+  reference: firestore.DocumentReference<T>,
+  data: T
+): Promise<void>;
+export function setDoc<T>(
+  reference: firestore.DocumentReference<T>,
+  data: Partial<T>,
+  options: firestore.SetOptions
+): Promise<void>;
+export function setDoc<T>(
+  reference: firestore.DocumentReference<T>,
+  data: T,
+  options?: firestore.SetOptions
+): Promise<void> {
+  const ref = cast<DocumentReference<T>>(reference, DocumentReference);
+  const firestore = cast(ref.firestore, Firestore);
+
+  const [convertedValue] = applyFirestoreDataConverter(
+    ref._converter,
+    data,
+    'setDoc'
+  );
+  const dataReader = newUserDataReader(firestore);
+  const parsed = dataReader.parseSetData(
+    'setDoc',
+    ref._key,
+    convertedValue,
+    options
+  );
+
+  return firestore
+    ._getFirestoreClient()
+    .then(firestoreClient =>
+      firestoreClient.write(parsed.toMutations(ref._key, Precondition.none()))
+    );
+}
+
+export function updateDoc(
+  reference: firestore.DocumentReference<unknown>,
+  data: firestore.UpdateData
+): Promise<void>;
+export function updateDoc(
+  reference: firestore.DocumentReference<unknown>,
+  field: string | firestore.FieldPath,
+  value: unknown,
+  ...moreFieldsAndValues: unknown[]
+): Promise<void>;
+export function updateDoc(
+  reference: firestore.DocumentReference<unknown>,
+  fieldOrUpdateData: string | firestore.FieldPath | firestore.UpdateData,
+  value?: unknown,
+  ...moreFieldsAndValues: unknown[]
+): Promise<void> {
+  const ref = cast<DocumentReference<unknown>>(reference, DocumentReference);
+  const firestore = cast(ref.firestore, Firestore);
+  const dataReader = newUserDataReader(firestore);
+
+  let parsed: ParsedUpdateData;
+  if (
+    typeof fieldOrUpdateData === 'string' ||
+    fieldOrUpdateData instanceof FieldPath
+  ) {
+    parsed = dataReader.parseUpdateVarargs(
+      'updateDoc',
+      ref._key,
+      fieldOrUpdateData,
+      value,
+      moreFieldsAndValues
+    );
+  } else {
+    parsed = dataReader.parseUpdateData(
+      'updateDoc',
+      ref._key,
+      fieldOrUpdateData
+    );
+  }
+
+  return firestore
+    ._getFirestoreClient()
+    .then(firestoreClient =>
```

Comment:
Same comment from previous PR's. Why is there chaining instead of just using async/await?

Reply:
Same response: )

"The outermost code needs to be async to ensure synchronous function execution, which impacts error propagation and execution order. The cast calls here, for example, would not throw but cause rejected Promises."

In this case, this also impact user input validation, which we want to throw as it is likely caused by invalid SDK usage.